src/Kvasir/Reconstitution/Lookups.cs

```diff
@@ -30,19 +30,22 @@ internal static partial class Lookup {
         public static object ByKey(DBData key, IEnumerable<object> entities, DataExtractionPlan extractor) {
             Debug.Assert(entities.All(e => e.GetType().IsInstanceOf(extractor.ExpectedSource)));
 
+            if (cacheDictionary_.TryGetValue(key, out object? result)) {
+                return result!;
+            }
+
             foreach (var entity in entities) {
-                var keyIdx = 0;
-                foreach (var value in extractor.ExecutePiecewise(entity)) {
-                    if (!value.Equals(key[keyIdx++])) {
-                        goto nextEntity;
-                    }
+                if (key.SequenceEqual(extractor.ExecutePiecewise(entity))) {
+                    cacheDictionary_[key] = entity;
+                    return entity;
                 }
-                return entity;
-                nextEntity: continue;
             }
 
             // This code should be unreachable
             throw new ApplicationException($"No entity found to match key: {string.Join(", ", key)}");
         }
+
+
+        private static readonly Dictionary<DBData, object> cacheDictionary_ = new();
```

Comment:
I think there's still a small problem here: what happens if two Entities of different types have the same key? I think we need to incorporate the source type of the passed-in extraction plan to account for that.

Reply:
Is that actually possible? If it is possible, that makes me think that you can have different entities of the same type with the same key, so we would need to fix that somehow too?

Reply:
Sure - Entities map to database tables, so the primary keys are only guaranteed to be unique within a set of Entities. Consider two Entities `Characters` and `Items` that both have an integer Field `ID` - you can only have one `Item` whose primary key is `[ { 7 } ]`, but that can _also_ be the primary key of a `Character`.

The rules of databases prevent two Entities of the same type from having the same primary key, so that's not something we have to actively guard against. The expectation is that whatever back-end provider being used (even if it's not technically a RDBMs) is going to enforce that.

Reply:
That's true - I just added type checking. All the unit tests pass, and I got to learn a lot about generic types!